src/simularium/VisData.ts

```diff
@@ -1,95 +1,47 @@
-import { compareTimes } from "../util";
+import { noop } from "lodash";
+
+import { nullCachedFrame } from "../util";
 
 import * as util from "./ThreadUtil";
-import {
-    AGENT_OBJECT_KEYS,
-    AgentData,
-    FrameData,
-    VisDataMessage,
-} from "./types";
-import { FrontEndError, ErrorLevel } from "./FrontEndError";
-import type { ParsedBundle } from "./VisDataParse";
+import { VisDataMessage, CachedFrame } from "./types";
 import { parseVisDataMessage } from "./VisDataParse";
-import { nullAgent } from "../constants";
+import { VisDataCache } from "./VisDataCache";
+import { ErrorLevel, FrontEndError } from "./FrontEndError";
 
 class VisData {
-    private frameCache: AgentData[][];
-    private frameDataCache: FrameData[];
-    private enableCache: boolean;
+    public frameCache: VisDataCache;
     private webWorker: Worker | null;
 
     private frameToWaitFor: number;
     private lockedForFrame: boolean;
-    private cacheFrame: number;
+
+    protected currentFrameNumber: number;
 
     public timeStepSize: number;
+    public onError: (error: FrontEndError) => void;
 
-    private static parseOneBinaryFrame(data: ArrayBuffer): ParsedBundle {
-        const parsedAgentDataArray: AgentData[][] = [];
-        const frameDataArray: FrameData[] = [];
+    private static parseOneBinaryFrame(data: ArrayBuffer): CachedFrame {
         const floatView = new Float32Array(data);
         const intView = new Uint32Array(data);
-        const parsedFrameData = {
-            time: floatView[1],
+
+        const frameData: CachedFrame = {
+            data: data,
             frameNumber: floatView[0],
+            time: floatView[1],
+            agentCount: intView[2],
+            size: data.byteLength + 8 * 4,
```

Comment:
This function has a much reduced role, simply constructs the cachedFrame object including the raw binary in the data field.